File: gateways/yarp/src/StackverseGateway/AccessTokenManager.cs
using System.Diagnostics;
using System.Globalization;
using System.Net;
using System.Text.Json;
using Microsoft.AspNetCore.Authentication;
using Microsoft.AspNetCore.Authentication.Cookies;
using Microsoft.AspNetCore.Authentication.OpenIdConnect;
using Microsoft.Extensions.Options;

namespace StackverseGateway;

/// <summary>
/// Hands out the session's access token for the /api relay, refreshing it against the
/// IdP token endpoint when it is about to expire and persisting the refreshed tokens
/// back into the Redis session.
///
/// This is a deliberate hand-rolled refresh path rather than a dependency on
/// Duende.AccessTokenManagement: the whole exchange is one form POST, and this repo
/// optimizes for self-contained code a reader can follow end to end. Concurrent
/// requests may occasionally refresh twice; Keycloak permits refresh-token reuse by
/// default, so the loser of the race simply stores an equally valid token.
/// </summary>
public sealed class AccessTokenManager(
    IOptionsMonitor<OpenIdConnectOptions> oidcOptions,
    RedisTicketStore ticketStore,
    ILogger<AccessTokenManager> logger)
{
    /// <summary>Refresh slightly early so a token cannot expire mid-flight to the backend.</summary>
    private static readonly TimeSpan ExpirySkew = TimeSpan.FromSeconds(30);

    /// <summary>
    /// Returns a valid access token for the authenticated session, or null when the
    /// session can no longer produce one (refresh token expired or revoked). Throws
    /// <see cref="IdpUnavailableException"/> when the IdP cannot be asked, or answers
    /// with its own failure instead of a verdict on the grant — either outcome says
    /// nothing about the session, so the caller must not destroy it.
    /// </summary>
    public async Task<string?> GetAccessTokenAsync(AuthenticateResult auth, CancellationToken cancellationToken)
    {
        var properties = auth.Properties!;
        var accessToken = properties.GetTokenValue("access_token");
        if (accessToken is null)
        {
            return null;
        }

        var expiresAt = properties.GetTokenValue("expires_at");
        if (expiresAt is not null
            && DateTimeOffset.TryParse(expiresAt, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var expiry)
            && expiry - ExpirySkew > DateTimeOffset.UtcNow)
        {
            return accessToken;
        }

        return await RefreshAsync(auth, cancellationToken);
    }

    private async Task<string?> RefreshAsync(AuthenticateResult auth, CancellationToken cancellationToken)
    {
        var properties = auth.Properties!;
        var refreshToken = properties.GetTokenValue("refresh_token");
        if (refreshToken is null)
        {
            return null;
        }

        var options = oidcOptions.Get(OpenIdConnectDefaults.AuthenticationScheme);

        // Only an IdP that *rejects* the refresh proves the session is dead. An IdP
        // that cannot be reached, answers 5xx/429, or answers garbage proves nothing,
        // so those paths must not fall through to "destroy the session": log the
        // dependency failure and throw, letting the /api guard fail the request
        // while the session stays.
        string newAccessToken, newRefreshToken;
        int expiresIn;
        var stopwatch = Stopwatch.StartNew();
        try
        {
            var metadata = await options.ConfigurationManager!.GetConfigurationAsync(cancellationToken);

            using var response = await options.Backchannel.PostAsync(
                metadata.TokenEndpoint,
                new FormUrlEncodedContent(new Dictionary<string, string>
                {
                    ["grant_type"] = "refresh_token",
                    ["refresh_token"] = refreshToken,
                    ["client_id"] = options.ClientId!,
                    ["client_secret"] = options.ClientSecret!,
                }),
                cancellationToken);

            if (!response.IsSuccessStatusCode)
            {
                // Only an authoritative rejection of the grant proves the session is
                // dead — RFC 6749 §5.2: a 400 (invalid_grant, expired/revoked refresh
                // token) or a 401 (client authentication). A 5xx or 429 is the IdP
                // failing at its own job and says nothing about the session.
                if (response.StatusCode is not (HttpStatusCode.BadRequest or HttpStatusCode.Unauthorized))
                {
                    logger.Event(LogLevel.Error, "dependency_call_failed", "failure",
                        $"Keycloak answered {(int)response.StatusCode} during token refresh; the session is kept",
                        fields:
                        [
                            ("dependency", "keycloak"),
                            ("duration_ms", stopwatch.ElapsedMilliseconds),
                            ("error_code", $"idp_status_{(int)response.StatusCode}"),
                        ]);
                    throw new IdpUnavailableException($"The IdP answered {(int)response.StatusCode} to the token refresh");
                }
                // degraded but self-healing (the caller destroys the session) — WARN per docs/LOGGING.md §5
                logger.Event(LogLevel.Warning, "token_refresh_failed", "failure",
                    $"Token refresh rejected by the IdP ({(int)response.StatusCode}); treating the session as expired",
                    fields: [("error_code", "idp_rejected"), ("idp_status", (int)response.StatusCode)]);
                return null;
            }

            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(cancellationToken));
            var root = payload.RootElement;
            newAccessToken = root.GetProperty("access_token").GetString()!;
            newRefreshToken = root.TryGetProperty("refresh_token", out var rotated) ? rotated.GetString()! : refreshToken;
            expiresIn = root.TryGetProperty("expires_in", out var lifetime) ? lifetime.GetInt32() : 300;
        }
        catch (IdpUnavailableException)
        {
            throw; // already logged at the throw site above
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw; // the client went away — not an IdP outage
        }
        catch (Exception exception)
        {
            logger.Event(LogLevel.Error, "dependency_call_failed",
                exception is OperationCanceledException ? "timeout" : "failure",
                "Keycloak was unreachable during token refresh; the session is kept",
                exception,
                ("dependency", "keycloak"),
                ("duration_ms", stopwatch.ElapsedMilliseconds),
                ("error_code", exception.GetType().Name));
            throw new IdpUnavailableException(exception);
        }

        properties.UpdateTokenValue("access_token", newAccessToken);
        properties.UpdateTokenValue("refresh_token", newRefreshToken);
        properties.UpdateTokenValue(
            "expires_at",
            DateTimeOffset.UtcNow.AddSeconds(expiresIn).ToString("o", CultureInfo.InvariantCulture));

        // Persist the rotated tokens so every gateway instance sees them.
        if (properties.Items.TryGetValue(RedisTicketStore.SessionKeyItem, out var sessionKey) && sessionKey is not null)
        {
            await ticketStore.RenewAsync(
                sessionKey,
                new AuthenticationTicket(auth.Principal!, properties, CookieAuthenticationDefaults.AuthenticationScheme));
        }

        return newAccessToken;
    }
}

/// <summary>
/// A token refresh failed because the IdP was unreachable, failing (5xx/429), or
/// unintelligible — a transient dependency outage, distinct from the IdP rejecting
/// the refresh token itself.
/// </summary>
public sealed class IdpUnavailableException : Exception
{
    public IdpUnavailableException(string message) : base(message)
    {
    }

    public IdpUnavailableException(Exception inner)
        : base("The IdP could not be reached to refresh the access token", inner)
    {
    }
}

File: gateways/spring-cloud-gateway/src/main/kotlin/dev/stackverse/gateway/relay/AccessTokenManager.kt
package dev.stackverse.gateway.relay

import dev.stackverse.gateway.common.logEvent
import org.slf4j.LoggerFactory
import org.slf4j.event.Level
import org.springframework.security.oauth2.client.OAuth2AuthorizeRequest
import org.springframework.security.oauth2.client.OAuth2AuthorizedClient
import org.springframework.security.oauth2.client.ReactiveOAuth2AuthorizedClientManager
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
import org.springframework.security.oauth2.client.web.server.ServerOAuth2AuthorizedClientRepository
import org.springframework.security.oauth2.core.OAuth2AuthorizationException
import org.springframework.security.oauth2.core.OAuth2ErrorCodes
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClientResponseException
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebSession
import reactor.core.publisher.Mono
import java.time.Duration
import java.time.Instant
import java.util.concurrent.TimeoutException

/**
 * Hands out the session's access token for the /api relay, refreshing it against the
 * IdP token endpoint when it is about to expire and persisting the refreshed tokens
 * back into the Redis-backed session.
 *
 * Spring Security owns the refresh grant exchange, response decoding, and rotated-token
 * persistence through [ReactiveOAuth2AuthorizedClientManager]. This adapter keeps only
 * the Stackverse-specific policy around the manager: an authoritative refresh-token
 * rejection destroys the session and relays anonymously, while IdP unavailability keeps
 * the session and lets the route answer 503. Concurrent requests may occasionally
 * refresh twice; Keycloak permits refresh-token reuse by default, so the loser of the
 * race simply stores an equally valid token.
 */
@Component
class AccessTokenManager(
    private val authorizedClients: ServerOAuth2AuthorizedClientRepository,
    private val authorizedClientManager: ReactiveOAuth2AuthorizedClientManager,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    /**
     * Emits a valid access token for the authenticated session, or completes empty
     * when the session can no longer produce one (refresh token expired or revoked) —
     * the session is destroyed and the caller relays anonymously. Errors with
     * [IdpUnavailableException] when the IdP cannot be asked, or answers with its own
     * failure instead of a verdict on the grant — either outcome says nothing about
     * the session, so it survives (docs/ARCHITECTURE.md).
     */
    fun accessToken(auth: OAuth2AuthenticationToken, exchange: ServerWebExchange): Mono<String> =
        authorizedClients
            .loadAuthorizedClient<OAuth2AuthorizedClient>(auth.authorizedClientRegistrationId, auth, exchange)
            // materialize absence: a plain switchIfEmpty would also fire when a
            // *rejected* refresh completes empty, destroying the session twice
            .singleOptional()
            .flatMap { maybeClient ->
                // an authenticated session without stored tokens cannot produce one
                val client = maybeClient.orElse(null) ?: return@flatMap destroyAndDegrade(auth, exchange)
                if (!isExpired(client)) {
                    Mono.just(client.accessToken.tokenValue)
                } else if (client.refreshToken == null) {
                    destroyAndDegrade(auth, exchange)
                } else {
                    refresh(auth, client, exchange)
                }
            }

    private fun refresh(
        auth: OAuth2AuthenticationToken,
        client: OAuth2AuthorizedClient,
        exchange: ServerWebExchange,
    ): Mono<String> {
        val start = System.nanoTime()
        fun elapsedMs() = Duration.ofNanos(System.nanoTime() - start).toMillis()

        val request = OAuth2AuthorizeRequest.withAuthorizedClient(client)
            .principal(auth)
            .attribute(ServerWebExchange::class.java.name, exchange)
            .build()

        return authorizedClientManager.authorize(request)
            .singleOptional()
            .flatMap { maybeRefreshed ->
                val refreshed = maybeRefreshed.orElse(null)
                    ?: return@flatMap destroyAndDegrade(auth, exchange)
                Mono.just(refreshed.accessToken.tokenValue)
            }
            .onErrorResume { e ->
                if (e.isRefreshGrantRejection()) {
                    // degraded but self-healing (the session is destroyed below) — WARN per docs/LOGGING.md §5
                    log.logEvent(
                        Level.WARN, "token_refresh_failed", "failure",
                        "Token refresh rejected by the IdP (${refreshRejectionCode(e)}); treating the session as expired",
                        "error_code" to "idp_rejected",
                        "oauth2_error" to refreshRejectionCode(e),
                    )
                    destroyAndDegrade(auth, exchange)
                } else {
                    // Anything that is not a verdict on the grant — unreachable,
                    // timeout, 5xx/429 OAuth error, or garbage in a success body —
                    // proves nothing about the session. Keep it and let the relay
                    // answer 503.
                    log.logEvent(
                        Level.ERROR, "dependency_call_failed",
                        if (e is TimeoutException) "timeout" else "failure",
                        "Keycloak was unavailable during token refresh; the session is kept",
                        "dependency" to "keycloak",
                        "duration_ms" to elapsedMs(),
                        "error_code" to dependencyErrorCode(e),
                        cause = e,
                    )
                    Mono.error(IdpUnavailableException("The IdP could not refresh the access token", e))
                }
            }
    }

    private fun isExpired(client: OAuth2AuthorizedClient): Boolean =
        client.accessToken.expiresAt?.minus(EXPIRY_SKEW)?.isAfter(Instant.now()) == false

    private fun Throwable.isRefreshGrantRejection(): Boolean =
        oauth2ErrorCode() == OAuth2ErrorCodes.INVALID_GRANT ||
            oauth2ErrorCode() == OAuth2ErrorCodes.INVALID_CLIENT ||
            causeChain().any { cause ->
                when (cause) {
                    is TokenEndpointResponseStatusException -> cause.statusCode.isRefreshGrantRejectionStatus()
                    is WebClientResponseException -> cause.statusCode.value().isRefreshGrantRejectionStatus()
                    else -> false
                }
            }

    private fun dependencyErrorCode(e: Throwable): String =
        if (e is OAuth2AuthorizationException) e.error.errorCode else e.javaClass.simpleName

    private fun refreshRejectionCode(e: Throwable): String =
        e.oauth2ErrorCode()
            ?: e.causeChain()
                .firstNotNullOfOrNull { cause ->
                    when (cause) {
                        is TokenEndpointResponseStatusException -> "http_${cause.statusCode}"
                        is WebClientResponseException -> "http_${cause.statusCode.value()}"
                        else -> null
                    }
                }
            ?: e.javaClass.simpleName

    private fun Throwable.oauth2ErrorCode(): String? =
        (this as? OAuth2AuthorizationException)?.error?.errorCode

    private fun Throwable.causeChain(): Sequence<Throwable> =
        generateSequence(this) { it.cause }

    private fun Int.isRefreshGrantRejectionStatus(): Boolean =
        this == 400 || this == 401

    /**
     * The session can no longer produce a token: destroy it and degrade the request
     * to anonymous. The SPA notices via the backend's 401 or /auth/session.
     */
    private fun destroyAndDegrade(auth: OAuth2AuthenticationToken, exchange: ServerWebExchange): Mono<String> =
        exchange.session
            .flatMap(WebSession::invalidate)
            .doOnSuccess {
                log.logEvent(
                    Level.INFO, "session_destroyed", "success",
                    "Session destroyed after a failed token refresh; request degraded to anonymous",
                    "reason" to "token_refresh_failed",
                    "actor" to auth.name,
                )
            }
            .then(Mono.empty())

    companion object {
        /** Refresh slightly early so a token cannot expire mid-flight to the backend. */
        private val EXPIRY_SKEW: Duration = Duration.ofSeconds(30)
    }
}

/**
 * A token refresh failed because the IdP was unreachable, failing (5xx/429), or
 * unintelligible — a transient dependency outage, distinct from the IdP rejecting
 * the refresh token itself.
 */
class IdpUnavailableException(message: String, cause: Throwable? = null) : RuntimeException(message, cause)
